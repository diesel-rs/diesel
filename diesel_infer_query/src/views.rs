use super::SchemaResolver;
use super::select::SelectField;
use super::select::infer_from_select;
use crate::error::Error;
use crate::error::Result;
use sqlparser::parser::ParserOptions;

/// An opaque representation of information
/// about a specific SQL view
///
/// Use the provided methods to access the information
#[derive(Debug, PartialEq)]
pub struct ViewData {
    pub(crate) fields: Vec<SelectField>,
}

impl ViewData {
    /// The number of fields returned by this VIEW
    pub fn field_count(&self) -> usize {
        self.fields.len()
    }

    /// Infer the nullablity of all VIEW fields
    ///
    /// This function returns a vector of optional booleans.
    /// The number and order of elements in this vector corresponds to
    /// the number and order of fields returned by this VIEW.
    ///
    /// Each value indicates whether the field is nullable (`Some(true)`),
    /// not nullable (`Some(false)`) or if the nullablity could
    /// not be inferred (`None`)
    ///
    /// This method accepts a generic [`SchemaResolver`]
    /// to query information about relations used in this
    /// view definition
    pub fn infer_nullability(
        &self,
        resolver: &mut dyn SchemaResolver,
    ) -> Result<Vec<Option<bool>>> {
        self.fields
            .iter()
            .map(|f| f.infer_nullability(resolver))
            .collect()
    }
}

/// Infer information about a given view definition
///
/// This method accepts both `CREATE VIEW xyz AS SELECT …` and
/// plain `SELECT …` statements as view definition.
pub fn parse_view_def(definition: &str) -> Result<ViewData> {
    let dialect = sqlparser::dialect::SQLiteDialect {};
    let options = ParserOptions::new();

    let stmt = sqlparser::parser::Parser::new(&dialect)
        .with_options(options)
        .try_with_sql(definition)?
        .parse_statement()?;

    let select = match stmt {
        sqlparser::ast::Statement::Query(query) => query,
        sqlparser::ast::Statement::CreateView { query, .. } => query,
        stmt => {
            return Err(Error::UnsupportedSql {
                msg: format!("Unexpected statement: `{stmt}`"),
            });
        }
    };
    let result = match &*select.body {
        sqlparser::ast::SetExpr::Select(select) => infer_from_select(select),
        // we likely want to support more complex queries here as well (UNION, CTE, etc)
        s => {
            return Err(Error::UnsupportedSql {
                msg: format!("Unsupported query kind: `{s}`"),
            });
        }
    };
    Ok(ViewData { fields: result? })
}
