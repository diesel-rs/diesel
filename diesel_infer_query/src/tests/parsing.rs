// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use crate::views;

#[track_caller]
pub(crate) fn check_parse_view(name: &'static str, def: &'static str) {
    let res = views::parse_view_def(def);
    assert!(
        res.is_ok(),
        "Failed to infer SQL with error: {}",
        res.unwrap_err()
    );
    let res = res.unwrap();

    let mut settions = insta::Settings::new();
    settions.set_raw_info(&(def.into()));
    settions.bind(|| {
        insta::assert_debug_snapshot!(name, res);
    });
}

#[test]
pub(crate) fn view_with_literals() {
    check_parse_view(
        "view_with_literals",
        "CREATE VIEW test AS SELECT 1 AS foo, 'foo' AS bar, NULL",
    );
}

#[test]
pub(crate) fn simple_table() {
    check_parse_view(
        "simple_table",
        "CREATE VIEW test AS SELECT users.id, name, hair_color as hair_colour FROM users",
    );
}

#[test]
pub(crate) fn simple_table_with_alias() {
    check_parse_view(
        "simple_table_with_alias",
        "CREATE VIEW test AS SELECT u.id, name, u.hair_color AS hair_colour FROM users as u",
    );
}
