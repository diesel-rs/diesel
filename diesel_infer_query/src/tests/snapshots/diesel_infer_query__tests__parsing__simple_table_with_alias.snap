---
source: diesel_infer_query/src/tests/parsing.rs
expression: res
info: "CREATE VIEW test AS SELECT u.id, name, u.hair_color AS hair_colour FROM users as u"
---
ViewData {
    fields: [
        SelectField {
            ident: None,
            kind: Field {
                schema: None,
                query_source: "users",
                field_name: "id",
                via_left_join: false,
            },
        },
        SelectField {
            ident: None,
            kind: Field {
                schema: None,
                query_source: "users",
                field_name: "name",
                via_left_join: false,
            },
        },
        SelectField {
            ident: Some(
                "hair_colour",
            ),
            kind: Field {
                schema: None,
                query_source: "users",
                field_name: "hair_color",
                via_left_join: false,
            },
        },
    ],
}
